tests/tools/oesign/test-engine/CMakeLists.txt

```diff
@@ -0,0 +1,97 @@
+# Copyright (c) Open Enclave SDK contributors.
+# Licensed under the MIT License.
+
+add_library(oesign_test_engine SHARED oesign-test-engine.c)
+
+target_include_directories(oesign_test_engine PRIVATE include)
+
+set(OESIGN_TEST_INPUTS_DIR ${CMAKE_CURRENT_BINARY_DIR}/../test-inputs)
+
+# Sign enclave
+add_custom_command(
+  OUTPUT oesign_test_enc.engine_signed
+  DEPENDS oesign_test_enc oesign_test_keys oesign_test_engine
+  COMMAND cmake -E copy $<TARGET_FILE:oesign_test_enc>
+          ${CMAKE_CURRENT_BINARY_DIR}
+  COMMAND
+    oesign sign -e oesign_test_enc -c ${OESIGN_TEST_INPUTS_DIR}/valid.conf -n
+    'oesign-test-engine' -p $<TARGET_FILE:oesign_test_engine> -i
+    ${OESIGN_TEST_INPUTS_DIR}/sign_key.private.pem
+  COMMAND cmake -E rename oesign_test_enc.signed oesign_test_enc.engine_signed)
+
+add_custom_target(oesign_engine_sign ALL DEPENDS oesign_test_enc.engine_signed)
+
+add_test(NAME tests/oesign-engine-verify-signed
+         COMMAND $<TARGET_FILE:oesign_test_host> oesign_test_enc.engine_signed)
+
+add_test(
+  NAME tests/oesign-engine-failed-engine-path
+  COMMAND
+    ${CMAKE_CURRENT_SOURCE_DIR}/oesign-test.py --oesign-binary
+    ${OE_BINDIR}/oesign --enclave-binary $<TARGET_FILE:oesign_test_enc>
+    --config-file ${OESIGN_TEST_INPUTS_DIR}/valid.conf --engine-id
+    'oesign-test-engine' --engine-load-path '/tmp/no_there' --key-id
+    ${OESIGN_TEST_INPUTS_DIR}/sign_key.private.pem --expect-fail)
+
+add_test(
+  NAME tests/oesign-engine-failed-engine-id
+  COMMAND
+    ${CMAKE_CURRENT_SOURCE_DIR}/oesign-test.py --oesign-binary
+    ${OE_BINDIR}/oesign --enclave-binary $<TARGET_FILE:oesign_test_enc>
+    --config-file ${OESIGN_TEST_INPUTS_DIR}/valid.conf --engine-id
+    'oesign-test' --engine-load-path $<TARGET_FILE:oesign_test_engine> --key-id
+    ${OESIGN_TEST_INPUTS_DIR}/sign_key.private.pem --expect-fail)
+
+add_test(
+  NAME tests/oesign-engine-failed-key-id
+  COMMAND
+    ${CMAKE_CURRENT_SOURCE_DIR}/oesign-test.py --oesign-binary
+    ${OE_BINDIR}/oesign --enclave-binary $<TARGET_FILE:oesign_test_enc>
+    --config-file ${OESIGN_TEST_INPUTS_DIR}/valid.conf --engine-id
+    'oesign-test-engine' --engine-load-path $<TARGET_FILE:oesign_test_engine>
+    --key-id 'bogus\ key' --expect-fail)
```

Comment:
This is more a `non existent key` test rather than an `invalid key` test.

Reply:
Not sure what you mean; the `--key-id` is a value only interpreted by the engine, i.e. as far as oesign is concerned, it's just an arbitrary string. The engine in this case is the mock test-engine, which chooses to interpret the `--key-id` value as a file path. oesign doesn't care about the distinction of how the engine fails on this, only that it is eventually informed that it is an invalid parameter.

Reply:
Ah Ok. I thought `oesign` was validating that argument.

Looking at the code currently, we don't seem to be reporting an invalid `--key-id` in a manner that the user can act upon. The openssl code-path seems to do this:
https://github.com/openenclave/openenclave/blob/11eede869d823fab6b25052480a40c35e34e94ce/host/crypto/openssl/key.c#L109-L110

Reply:
Yeah, I agree that the engine handling code path is not ideal in terms of its parameter handling; there should probably be a pass to standardize the parameter and error handling behavior in oesign (and more generally related tools like oesgx) but that's out of scope for this series of PRs.